### PR TITLE
feat(gateway): validate GatewayInstance resources 

### LINKED ISSUE #3502: Validate `GatewayInstance` CRD
### Description

* Validate `Tags` with the tag and value regexes, and requiring a service tag.

See originally https://github.com/kumahq/kuma/pull/3469#discussion_r767336305

### PR DESCRIPTION
**Pull request by michaelbeaumont**
### Summary

Closes #3502

### CODE REVIEW

**Review by jpeach — CHANGES_REQUESTED** [review]

File: pkg/core/resources/apis/mesh/validators.go [hunk 1]
```diff
@@ -278,3 +278,20 @@ func SelectorKeyNotInSet(keyName ...string) TagKeyValidatorFunc {
 			return err
 		})
 }
+
+// ValidateTags checks whether the tags and their values are valid.
+func ValidateTags(tags map[string]string) validators.ValidationError {
```
[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply pkg/core/resources/apis/mesh/validators.go:283]: https://github.com/kumahq/kuma/pull/3580

File: pkg/core/resources/apis/mesh/zone_ingress_validator.go [hunk 2]
```diff
@@ -43,7 +43,7 @@ func (r *ZoneIngressResource) validateAvailableServices(path validators.PathBuil
 		if _, ok := availableService.Tags[mesh_proto.ServiceTag]; !ok {
 			err.AddViolationAt(p.Field("tags").Key(mesh_proto.ServiceTag), "cannot be empty")
 		}
-		err.AddErrorAt(p.Field("tags"), validateTags(availableService.GetTags()))
+		err.AddErrorAt(p.Field("tags"), ValidateTags(availableService.GetTags()))
```
[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply pkg/core/resources/apis/mesh/zone_ingress_validator.go]: fixed in `master`

File: pkg/core/resources/apis/mesh/validators.go [hunk 3]
```diff
@@ -278,3 +278,20 @@ func SelectorKeyNotInSet(keyName ...string) TagKeyValidatorFunc {
...
+	var result validators.ValidationError
+	for name, value := range tags {
+		if value == "" {
+			result.AddViolationAt(validators.RootedAt("tags").Key(name), `tag value cannot be empty`)
```
[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply pkg/core/resources/apis/mesh/validators.go]: fixed in `master` by #3580

[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply pkg/core/resources/apis/mesh/validators.go]: fixed in `master` by #3580

File: pkg/plugins/runtime/k8s/webhooks/gateway_instance_validator.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	}
+
+	return admission.Allowed("")
+}
+
+func (h *GatewayInstanceValidator) ValidateUpdate(ctx context.Context, req admission.Request) admission.Response {
+	gatewayInstance := &mesh_k8s.GatewayInstance{}
+	if err := h.decoder.DecodeRaw(req.Object, gatewayInstance); err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
```
Comment (jpeach) [inline pkg/plugins/runtime/k8s/webhooks/gateway_instance_validator.go:69]: IIUC can just use `Decode` here. I'm not really sure why other validators want to validate both the old and the new objects - unless there's a possibility of making an invalid state transition, I would expect that only the new object matters.

**Review by jpeach — APPROVED** [review]

**Pull request merged**